### PR TITLE
Allow creating snapshot from VM snapshot

### PR DESCRIPTION
**Pull request by ravening**
### Description
If `kvm.snapshot.enabled` is set to false then we cant create snapshot from
VM snapshot. With this change, its possible to create snapshot from VM
snapshot even when the global setting is set to false.
Note that you still cant directly create a snapshot from volume though

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
I am struggling to see the use-case. Can you explain why you want this, @ravening?

**Comment by ravening** [comment]
> I am struggling to see the use-case. Can you explain why you want this, @ravening?


@DaanHoogland 

sometimes we need to create a template or volume from snapshot. If the global setting is false then we cant create snapshot and hence we cant create template also. But with this change, we are able to create those

**Comment by DaanHoogland** [comment]
so if you want to create a template from a vm-snapshot you need to first create a volume snapshot from it, tnx @ravening .

**Comment by weizhouapache** [comment]
> I am struggling to see the use-case. Can you explain why you want this, @ravening?

for stopped vms, we can create volume snapshot no matter if kvm.snapshot.enabled is true or false.

for running vms, if kvm.snapshot.enabled is false, we cannot create volume snapshot
if kvm.snapshot.enabled is true, we can create volume snapshot by following steps in cloudstack if image is qcow2 format.
(1) create vm snapshot (vm will be paused)
(2) create volume snapshot from the vm snapshot

If we already have a vm snapshot, even if kvm.snapshot.enabled is false, we should be able to create volume snapshot from vmsnapshot (step 2 above), as vm will not be paused and vm image will not be touched.

**Comment by rhtyd** [comment]
@weizhouapache @ravening any update on this?

**Comment by ravening** [comment]
> @weizhouapache @ravening any update on this?

not sure what update is needed here @rhtyd

**Comment by DaanHoogland** [comment]
me neither @rhtyd are you missing something here?

**Review by DaanHoogland — APPROVED** [review]
code looks good, any marvin or other test automation @ravening ?

**Comment by ravening** [comment]
> code looks good, any marvin or other test automation @ravening ?

@DaanHoogland didnt think of writing any automation test cases.. tested it manually

**Comment by DaanHoogland** [comment]
loads of travis errors @ravening , can you have a look?

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@ravening pl check snapshot related failures

**Comment by rhtyd** [comment]
Re-ping @ravening

**Comment by ravening** [comment]
> Re-ping @ravening

@rhtyd will look into it in few days.. kind of busy with some other work

**Comment by rhtyd** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -2692,7 +2692,7 @@ public Snapshot allocSnapshot(Long volumeId, Long policyId, String snapshotName,
             throw new InvalidParameterValueException("VolumeId: " + volumeId + " please attach this volume to a VM before create snapshot for it");
         }
 
-        return snapshotMgr.allocSnapshot(volumeId, policyId, snapshotName, locationType);
+        return snapshotMgr.allocSnapshot(volumeId, policyId, snapshotName, locationType, null);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2695]: @ravening null here means true or false ? if false, you can keep the old method.

```suggestion
        return snapshotMgr.allocSnapshot(volumeId, policyId, snapshotName, locationType, null);
```

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java [hunk 2]
```diff
@@ -1049,7 +1049,7 @@ private Type getSnapshotType(IntervalType intvType) {
         return null;
     }
 
-    private boolean hostSupportSnapsthotForVolume(HostVO host, VolumeInfo volume) {
+    private boolean hostSupportSnapsthotForVolume(HostVO host, VolumeInfo volume, boolean isFromVmSnapshot) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:1052]: can you correct the type error in method name

```suggestion
    private boolean hostSupportsSnapshotForVolume(HostVO host, VolumeInfo volume, boolean isFromVmSnapshot) {
```

Comment (weizhouapache) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:1052]: good suggestion. @ravening could you change it ?

**Comment by rhtyd** [comment]
Ping @ravening

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java (hunk 2 above)
Comment (ravening) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: @sureshanaparti made the change. can you review again?

**Comment by ravening** [comment]
@Pearl1594 if possible can you run the test again?

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by ravening** [comment]
@DaanHoogland @weizhouapache ran the tests again in my setup and these didnt fail

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by nvazquez — APPROVED** [review]

**Pull request merged**